### PR TITLE
[Feat] Improved data table rendering

### PR DESCRIPTION
**Pull request by heshan0131**
- Use canvas to calculate cell size based on data
- Add functionality to pin columns
Signed-off-by: Shan He <heshan0131@gmail.com>

### CODE REVIEW

**Comment by heshan0131** [comment]
#### 1. Calculate cell size based on content, expand cells by title or row content to fix window size
Screen Shot 2020-03-17 at 10 56 38 PM

#### 2. Added action to sort and pin cells
ezgif com-video-to-gif (1)

File: src/components/common/data-table/button.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  display: flex;
+  align-items: center;
+  padding: 0;
+
+  &:hover {
+    opacity: 0.8;
+  }
+`;
+
+class Button extends Component {
```
Comment (macrigiuseppe) [inline src/components/common/data-table/button.js]: why don't we use function components for new react elements?

File: src/components/common/item-selector/dropdown-list.js [hunk 2]
```diff
@@ -37,9 +37,10 @@ export const ListItem = ({value, displayOption = defaultDisplay}) => (
 );
 
 const DropdownListWrapper = styled.div`
-  background-color: ${props => props.theme.dropdownListBgd};
+  background-color: ${props =>
+    props.light ? props.theme.dropdownListBgdLT : props.theme.dropdownListBgd};
   border-top: 1px solid ${props => props.theme.dropdownListBorderTop};
-  ${props => props.theme.dropdownList};
+  ${props => (props.light ? props.theme.dropdownListLT : props.theme.dropdownList)};
```
Comment (macrigiuseppe) [inline src/components/common/item-selector/dropdown-list.js]: shouldn't the theme be in charge of providing the correct color without having kepler checking for __light__?

File: src/components/kepler-gl.js [hunk 3]
```diff
@@ -333,67 +335,71 @@ function KeplerGlFactory(
...
-              <PlotContainer
-                width={width}
-                height={height}
-                exportImageSetting={uiState.exportImage}
-                mapFields={mapFields}
-                addNotification={uiStateActions.addNotification}
-                startExportingImage={uiStateActions.startExportingImage}
-                setExportImageDataUri={uiStateActions.setExportImageDataUri}
-                setExportImageError={uiStateActions.setExportImageError}
+        <RootContext.Provider value={this.root}>
```
Comment (macrigiuseppe) [inline src/components/kepler-gl.js]: why are we passing this context?
is to just support the portaled component?

File: src/components/side-panel/layer-panel/custom-palette.js [hunk 4]
```diff
@@ -21,10 +21,9 @@
 import React, {Component, createRef} from 'react';
 import classnames from 'classnames';
 import PropTypes from 'prop-types';
-import {createSelector} from 'reselect';
-import styled, {withTheme, css} from 'styled-components';
+import styled, {css} from 'styled-components';
 import {sortableContainer, sortableElement, sortableHandle} from 'react-sortable-hoc';
-import Modal from 'react-modal';
+import Portaled from '../../common/portaled';
```
Comment (macrigiuseppe) [inline src/components/side-panel/layer-panel/custom-palette.js]: from 'components/common/portaled'

File: src/components/kepler-gl.js (hunk 3 above)
Comment (heshan0131) [reply src/components/kepler-gl.js]: yes, so all the modal from Portabled will be mounted under kepler.gl component

File: src/components/common/item-selector/dropdown-list.js (hunk 2 above)
Comment (heshan0131) [reply src/components/common/item-selector/dropdown-list.js]: well yes and no, this is the dropdown inside a modal, and all models are having the light theme. so we need a way to render the dropdown differently

File: src/components/common/data-table/cell-size.js [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+ * Measure rows and column content to determin min width for each column
+ * @param {*} param0
+ */
+export function renderedSize({
+  text: {rows, column},
+  type = 'string',
+  colIdx,
+  numRowsToCalculate = 10,
+  fontSize = 12,
+  font = 'Lato',
```
[the remark this answers is not in the document]
Comment (heshan0131) [reply src/components/common/data-table/cell-size.js]: that's the default canvas font

[the remark this answers is not in the document]
Comment (heshan0131) [reply src/components/common/data-table/cell-size.js]: default canvas font

File: src/components/common/data-table/canvas.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+    Except... Because scrolling is managed by a separate thread, and JavaScript
+    is only periodically notified of the updated position, there's some latency
+    issues with this.
+
+    We can hack around this by using a niche property of canvas that removes the
+    delay in scroll event firing! Easiest way to reproduce: enable "Trace React updates"
+    in React DevTools (it works by overlaying a viewport-wide canvas over the document).
+  */
+export default styled.canvas`
```
[the remark this answers is not in the document]
Comment (heshan0131) [reply src/components/common/data-table/canvas.js]: ok

**Pull request merged**